### ISSUE TITLE
Differentiable cosmology pipeline

### ISSUE THREAD

**Issue by sambit-giri** [sambit-giri/tools_cosmo#5]
## Motivation

Making the pipeline fully differentiable would enable:
- Gradient-based inference (HMC, variational inference) without finite-difference approximations
- Automatic differentiation through the entire forward model (P(k) → observables)
- Integration with JAX/PyTorch-based ML frameworks for emulation and parameter estimation

### 1. Differentiable Boltzmann solver — DISCO-EB
Replace or augment CAMB/CLASS with [DISCO-EB](https://github.com/hsimonfroy/DISCOeb), a JAX-based differentiable Boltzmann solver.
- Provides `jax.grad`-compatible outputs for the linear power spectrum
- Drop-in for the `param.cosmo.solver = 'camb'` path — add a `'disco-eb'` option in `run_BoltzmannSolver.py` and `cosmo.py`
- Requires updating `Ez_model`, `comoving_distance`, `distance_modulus`, etc. to accept JAX arrays

### 2. Differentiable non-linear P(k) — custom halo model
HMcode (and pyHMcode) are not differentiable. A native JAX halo model implementation is needed:
- Halo mass function: differentiable Press-Schechter / Sheth-Tormen (already in `massfct.py` — port to JAX)
- Halo profile: NFW in JAX with differentiable concentration-mass relation
- 1-halo + 2-halo terms assembled in JAX
- Validate against HMcode-2020 outputs

### 3. Differentiable distance and growth factor
- `Omega_DE`, `Ez_model`, `comoving_distance` — replace `scipy.integrate.quad` with JAX-compatible ODE solvers (`diffrax`)
- Growth factor ODE solver in `growth_factor_solveODE` — port to `diffrax`

### 4. End-to-end gradient flow
Once 1–3 are in place, the chain
```
cosmological params → DISCO-EB → P_lin(k) → halo model → P_nl(k) → observables
```
should be fully `jax.grad`-differentiable, enabling Fisher matrix computation and HMC sampling at negligible extra cost.

## Dependencies to add
- `jax`, `jaxlib`
- `diffrax` (JAX ODE solver)
- `disco-eb` (JAX Boltzmann solver)

These should be optional extras to avoid breaking the existing CAMB/CLASS workflow:
```toml
[project.optional-dependencies]
jax = ["jax", "jaxlib", "diffrax"]
```

## References
- DISCO-EB: https://github.com/hsimonfroy/DISCOeb
- diffrax: https://github.com/patrick-kidger/diffrax
- Halo model review: Cooray & Sheth 2002 (astro-ph/0206508)
- HMcode-2020: Mead et al. 2021 (arXiv:2009.01858)
